Openbsd rcctl

jarmani opened the pull request:
Work to improve the OpenBsdService service backend.

It uses rcctl on recent OpenBSD to enable / disable services. 
The behavior is not changed on older OpenBSD version.

Thanks to @eest who did a lot of the work and ajacoutot@openbsd.org who helped and improved  rcctl to fit your needs.

jarmani committed:
Add enable / disable of services for OpenBSD if rcctl is present

eest committed:
Make "enabled" code aware of --check mode.

eest committed:
Check rc and print error message if any.

It is probably good to use stdout before printing a generic error
message as well.

eest committed:
Fail if "rcctl status" writes to stderr.

eest committed:
Tweak error checking for "enabled" code.

Based on input from @jarmani:
* A return value of 2 now means a service does not exist. Instead of
  trying to handle the different meanings of rc after running "status",
  just look at stderr to know if something failed.
* Skip looking at stdout to make the code cleaner. Any errors should
  turn up on stderr.

eest committed:
Depend more on rcctl if it is present.

* Make the module support enable/disable of special services like pf via rcctl.
  Idea and method from @jarmani.
* Make the module handle when the user supplied 'arguments' variable does not
  match the current flags in rc.conf.local.
* Update description now that the code tries to use rcctl for everything if it
  is available.

eest committed:
Simplify self.arguments logic.

Strange logic pointed out by @jarmani, thanks!

eest committed:
Multiple fixes for OpenBSD rcctl handling.

* Use the newly added 'default' argument to know if the default flags are set
  or not.
* Handle that 'status' may either return flags or YES/NO.
* Centralize flag handling logic.
* Set action variable after check if we need to keep going.

Big thanks to @ajacoutot for implementing the rcctl 'default' argument.

jarmani committed:
Merge pull request #2 from eest/openbsd_rcctl

Multiple fixes for OpenBSD rcctl handling.

eest commented:
For whatever my word is worth as the original author of the OpenBSD bits: I am OK with this going in.

gangelop commented:
Yes, I want this! :+1:

gangelop commented:
~~If I understand this correctly, this also fixes the status checking for the service module (#200), which will also work on OpenBSD 5.5 (which doesn't have rcctl yet). Is that correct?~~

I'm wrong. Nothing to see here...

eest commented:
@lathan: I might be missing something, but I am not sure what you think is broken. Please see my comment in #200.

eest committed:
Fix typo: Replace "N0" with "NO".

jarmani committed:
Merge pull request #3 from eest/openbsd_rcctl

Fix typo: Replace "N0" with "NO".

eest committed:
Handle string returned by 'default' correctly.

We need to handle the string returned by 'default' in the same way we handle
the string returned by 'status' since the resulting flags are compared later.

jarmani committed:
Merge pull request #4 from eest/openbsd_rcctl

Handle string returned by 'default' correctly.

nmeum commented:
Looks good and seems to work. :+1:
